core/domain/platform_parameter_domain.py

```
@@ -385,18 +454,25 @@ def _match_version_expression(self, op, value, client_version):
         is_client_version_larger = self._is_first_version_smaller(
             value, client_version_without_hash
         )
+        matched = False
```

Should we rather `return False` at the end instead of creating a variable? Or maybe should we throw an error if the operation does not exist?

---

ditto below

---

If we `return false` at the end rather than creating a variable then we have total of 7 return statements in the function which exceeds the max limit of return statements ( which is 6 ).

Also, method `_match_version_expression` and `_match_version_flavor` are only called in `_evaluate_single_value`. And in `_evaluate_single_value` we are already checking if operator is correct or not. https://github.com/oppia/oppia/blob/ce25d73163d4a084bce5232fe35bb258ea8daba5/core/domain/platform_parameter_domain.py#L274
So, if we want to raise error for incorrect operator in  `_match_version_expression` and `_match_version_flavor` methods, then that error is never called/raised, because we are already checking operator in  `_evaluate_single_value`

---

Hmm I think we can remove that one and throw some here, or throw them in both places.

---

In `_evaluate_single_value`, we are checking all the cases of operators with their corresponding types.
https://github.com/oppia/oppia/blob/ba4f072e494fd59df53fecc37e67cea7f9727234/core/domain/platform_parameter_domain.py#L274
But if remove this check from `_evaluate_single_value` and move it to `_match_version_expression` and `_match_version_flavor`, then there we can only checking if the type is `app_version_flavor` or `app_version` and the operator belongs to list [=, <=, >=, <, >] or not, and by doing this we lose the cases where:
1. type is `server_mode` and operator is `=`.
2. type is `platform_type` and operator is `=`.
3. type is `browser_type` and operator is `=`.

---

Okay, can we then throw in both places?

---

Ok, Done!